### PR TITLE
Add addgroup function

### PR DESCRIPTION
**Pull request by maorJsonar**

### CODE REVIEW

File: grp.py [hunk 1]
```diff
@@ -26,3 +26,8 @@ def getgrnam(self, name):
 
     def getgrall(self):
         return self.db
+
+    def addgroup(self, name, gid=None):
```
Comment (oorijsonar) [inline grp.py:30]: Rename to `groupadd`. Like Linux!

File: grp.py [hunk 2]
```diff
@@ -26,3 +26,8 @@ def getgrnam(self, name):
...
+        if not gid:
+            gid = self.db[-1].gr_gid + 1
```
Comment (oorijsonar) [inline grp.py:32]: I don't get this. Why is this the value you're assigning to the `gid`?

File: grp.py (hunk 1 above)
Comment (oorijsonar) [inline grp.py:30]: Should this function be okay with any input and not raise exceptions?

**Review by oorijsonar — CHANGES_REQUESTED** [review]

File: grp.py (hunk 2 above)
Comment (maorJsonar) [reply grp.py:32]: If the user doesn't provide the gid I just thought to take the last gid and increase it by 1.

Comment (oorijsonar) [reply grp.py:32]: That might be taken though, no?

File: grp.py (hunk 1 above)
Comment (maorJsonar) [reply grp.py]: Added some validation.

File: grp.py [hunk 3]
```diff
@@ -26,3 +28,22 @@ def getgrnam(self, name):
 
     def getgrall(self):
         return self.db
+
+    def groupadd(self, args):
```
Comment (oorijsonar) [inline grp.py:34]: You can't raise random exceptions in the mock. You have to see what kind of exception the mocked function would raise in these cases (my guess is `CalledProcessError`).

**Review by oorijsonar — CHANGES_REQUESTED** [review]

File: grp.py (hunk 3 above)
Comment (maorJsonar) [reply grp.py]: Done

File: grp.py [hunk 4]
```diff
@@ -26,3 +28,20 @@ def getgrnam(self, name):
...
+
+        for group in self.db:
+            if gid is not None and int(gid) == group.gr_gid:
+                raise CalledProcessError(f"groupadd: GID '{gid}' already exists")
+
+            if name == group.gr_name:
+                raise CalledProcessError(f"groupadd: group '{name}' already exists")
+
+        if gid is None:
+            gid = max([group.gr_gid for group in self.db])
```
Comment (oorijsonar) [inline grp.py:45]: Will this work? The max is already there isn't it?

Comment (maorJsonar) [reply grp.py]: Nop, you are right, sorry its probably the hour.

**Review by oorijsonar — APPROVED** [review]
Does the `CalledProcessError` really have that as a message?

**Comment by maorJsonar** [comment]
> Does the `CalledProcessError` really have that as a message?

Yes, that's what I got trying to run it with the original `process.run()`

**Pull request merged**